File: python/pyautd3/modulation/square.py
"""
File: square.py
Project: modulation
Created Date: 14/09/2023
Author: Shun Suzuki
-----
Last Modified: 10/10/2023
Modified By: Shun Suzuki (suzuki@hapis.k.u-tokyo.ac.jp)
-----
Copyright (c) 2023 Shun Suzuki. All rights reserved.

"""


from pyautd3.emit_intensity import EmitIntensity
from pyautd3.internal.modulation import IModulationWithSamplingConfig
from pyautd3.native_methods.autd3capi import NativeMethods as Base
from pyautd3.native_methods.autd3capi import SamplingMode
from pyautd3.native_methods.autd3capi_def import ModulationPtr


class Square(IModulationWithSamplingConfig):
    """Square wave modulation."""

    _freq: float
    _low: EmitIntensity | None
    _high: EmitIntensity | None
    _duty: float | None
    _mode: SamplingMode | None

    def __init__(self: "Square", freq: float) -> None:
        """Constructor.

        Arguments:
        ---------
            freq: Frequency (Hz)
        """
        super().__init__()
        self._freq = freq
        self._low = None
        self._high = None
        self._duty = None
        self._mode = None

    def with_low(self: "Square", low: int | EmitIntensity) -> "Square":
        """Set low level intensity.

        Arguments:
        ---------
            low: Low level intensity
        """
        self._low = EmitIntensity._cast(low)
        return self

    def with_high(self: "Square", high: int | EmitIntensity) -> "Square":
        """Set high level intensity.

        Arguments:
        ---------
            high: High level intensity
        """
        self._high = EmitIntensity._cast(high)
        return self

    def with_duty(self: "Square", duty: float) -> "Square":
        """Set duty ratio which is defined as `Th / (Th + Tl)`, where `Th` is high level duration and `Tl` is low level duration.

        Arguments:
        ---------
            duty: Duty ratio (from 0 to 1)
        """
        self._duty = duty
        return self

    def with_mode(self: "Square", mode: SamplingMode) -> "Square":
        """Set sampling mode.

        Arguments:
        ---------
            mode: Sampling mode
        """
        self._mode = mode
        return self

    def _modulation_ptr(self: "Square") -> ModulationPtr:
        ptr = Base().modulation_square(self._freq)
        if self._low is not None:
            ptr = Base().modulation_square_with_low(ptr, self._low.value)
        if self._high is not None:
            ptr = Base().modulation_square_with_high(ptr, self._high.value)
        if self._duty is not None:
            ptr = Base().modulation_square_with_duty(ptr, self._duty)
        if self._config is not None:
            ptr = Base().modulation_square_with_sampling_config(ptr, self._config._internal)
        if self._mode is not None:
            ptr = Base().modulation_square_with_mode(ptr, self._mode)
        return ptr

File: dotnet/cs/src/Modulation/Square.cs
/*
 * File: Square.cs
 * Project: Modulation
 * Created Date: 13/09/2023
 * Author: Shun Suzuki
 * -----
 * Last Modified: 08/12/2023
 * Modified By: Shun Suzuki (suzuki@hapis.k.u-tokyo.ac.jp)
 * -----
 * Copyright (c) 2023 Shun Suzuki. All rights reserved.
 * 
 */

#if UNITY_2018_3_OR_NEWER
#define USE_SINGLE
#endif

#if UNITY_2020_2_OR_NEWER
#nullable enable
#endif

using AUTD3Sharp.NativeMethods;

#if USE_SINGLE
using float_t = System.Single;
#else
using float_t = System.Double;
#endif

namespace AUTD3Sharp.Modulation
{
    /// <summary>
    /// Square wave modulation
    /// </summary>
    public sealed class Square : Internal.ModulationWithSamplingConfig<Square>
    {
        private readonly float_t _freq;
        private EmitIntensity? _low;
        private EmitIntensity? _high;
        private float_t? _duty;
        private SamplingMode? _mode;

        public Square(float_t freq)
        {
            _freq = freq;
            _low = null;
            _high = null;
            _duty = null;
            _mode = null;
        }

        /// <summary>
        /// Set low level amplitude
        /// </summary>
        /// <param name="low">low level intensity</param>
        /// <returns></returns>
        public Square WithLow(byte low)
        {
            _low = new EmitIntensity(low);
            return this;
        }

        /// <summary>
        /// Set low level amplitude
        /// </summary>
        /// <param name="low">low level intensity</param>
        /// <returns></returns>
        public Square WithLow(EmitIntensity low)
        {
            _low = low;
            return this;
        }

        /// <summary>
        /// Set high level amplitude
        /// </summary>
        /// <param name="high">high level intensity</param>
        /// <returns></returns>
        public Square WithHigh(byte high)
        {
            _high = new EmitIntensity(high);
            return this;
        }

        /// <summary>
        /// Set high level amplitude
        /// </summary>
        /// <param name="high">high level intensity</param>
        /// <returns></returns>
        public Square WithHigh(EmitIntensity high)
        {
            _high = high;
            return this;
        }

        /// <summary>
        /// Set duty ratio
        /// </summary>
        /// <remarks>Duty ratio is defined as `Th / (Th + Tl)`, where `Th` is high level duration, and `Tl` is low level duration.</remarks>
        /// <param name="duty">normalized amplitude (0.0 - 1.0)</param>
        /// <returns></returns>
        public Square WithDuty(float_t duty)
        {
            _duty = duty;
            return this;
        }

        /// <summary>
        /// Set sampling mode
        /// </summary>
        /// <param name="mode">sampling mode</param>
        /// <returns></returns>
        public Square WithMode(SamplingMode mode)
        {
            _mode = mode;
            return this;
        }

        internal override ModulationPtr ModulationPtr()
        {
            var ptr = NativeMethodsBase.AUTDModulationSquare(_freq);
            if (_low != null)
                ptr = NativeMethodsBase.AUTDModulationSquareWithLow(ptr, _low.Value.Value);
            if (_high != null)
                ptr = NativeMethodsBase.AUTDModulationSquareWithHigh(ptr, _high.Value.Value);
            if (_duty != null)
                ptr = NativeMethodsBase.AUTDModulationSquareWithDuty(ptr, _duty.Value);
            if (Config != null)
                ptr = NativeMethodsBase.AUTDModulationSquareWithSamplingConfig(ptr, Config.Value.Internal);
            if (_mode != null)
                ptr = NativeMethodsBase.AUTDModulationSquareWithMode(ptr, _mode.Value);
            return ptr;
        }
    }
}

#if UNITY_2020_2_OR_NEWER
#nullable restore
#endif
